docs/design/factories.md

```diff
@@ -0,0 +1,234 @@
+populate them:
+
+```go
+resource := framework.Resource{
+	Schema: map[string]*Schema{},
+	Create: createFunc,
+	Read: readFunc,
+	Update: updateFunc,
+	Delete: deleteFunc,
+}
+
+newResource := framework.Resource{}
+newResource.Schema = resource.Schema
+newResource.Create = resource.Create
+newResource.Read = resource.Read
+newResource.Update = resource.Update
+newResource.Delete = resource.Delete
+```
+
+we can use this in a helper, to do something like:
+
+```go
+type Provider struct {
+	Resources map[string]Resource
+}
+```
+
+and do the instantiation ourselves behind the scenes.
+
+This works only if the type is defined by the framework. Types defined by the
+provider would need to use reflection to achieve this outcome.
+
+### Factory types
+
+Instead of using functions, we can have an interface for a factory type that
+providers define:
+
+```go
+type ResourceFactory interface {
+  NewResource() Resource
+}
+```
+
+Which can then be used instead of the type the factory will return when asking
+users for resources, data sources, and providers:
+
+```go
+type Provider struct {
+  Resources map[string]ResourceFactory
+}
+```
+
+This works whether the type is defined by the provider (the `NewResource()`
+method would return an interface) or by the framework.
+
+### Separating types and values
+
+The SDK currently conflates two separate ideas: the type of a resource--i.e.,
+the shape a resource takes, the general idea of a resource--and the value of a
+resource--e.g. a specific instance of that resource, a single resource
+definition in state.
+
+Rather than conflating types and values, we can separate them out into two
+different Go implementations. The type can then serve as a factory that
+contains information common to values, and values can surface the
+implementations that are used for specific instances of a type:
+
+```go
+type ResourceType interface {
+  GetSchema() *tfprotov5.Schema
+  NewValue() ResourceValue
+}
+
+type ResourceValue interface {
+  Create(ctx context.Context, *CreateResourceRequest, *CreateResourceResponse)
+  Read(ctx context.Context, *ReadResourceRequest, *ReadResourceResponse)
+  Update(ctx context.Context, *UpdateResourceRequest, *UpdateResourceResponse)
+  Delete(ctx context.Context, *DeleteResourceRequest, *DeleteResourceResponse)
+}
+```
+
+We can then use types to instantiate new values at runtime:
+
+```go
+type Provider struct {
+  Resources map[string]ResourceType
+}
+```
+
+This works whether the type is defined by the provider (the `NewValue()` method
+would return an interface) or by the framework.
+
+## Trade-offs
+
+### Mutability of data
+
+As a general rule of thumb, resources, data sources, and providers should not change at runtime. It's hard to imagine a scenario where modifying the schema, validation, or CRUD implementations while the server is running is a good idea, and we should consider that scenario out of scope for this design.
```

Comment:
I think, without work in Terraform core, we should strongly advise providers away from this. @alexsomesan may have more information, but I think Terraform will be _very_ unhappy if you change the schema of anything at runtime.

Also, GetSchema RPCs currently happen before ConfigureProvider RPCs, I believe.

Reply:
Yes, good point. If we see a provider that calculates the response to `GetSchema` based on (say) an API response as "changing the schema at runtime" or "while the server is running", then I think it is a plausible design, but since, as you rightly point out, this must happen before `ConfigureProvider`, I think it's sufficiently far from our intended use case as to be considered out of scope for the framework design.